mrmustard/lab_dev/circuit_components.py

```diff
@@ -47,40 +47,55 @@ class CircuitComponent:
 
     def __init__(
         self,
-        name: str,
+        name: Optional[str] = None,
         representation: Optional[Representation] = None,
         modes_out_bra: Optional[Sequence[int]] = None,
         modes_in_bra: Optional[Sequence[int]] = None,
         modes_out_ket: Optional[Sequence[int]] = None,
         modes_in_ket: Optional[Sequence[int]] = None,
     ) -> None:
-        self._name = name
+        self._name = name or ""
         self._wires = Wires(modes_out_bra, modes_in_bra, modes_out_ket, modes_in_ket)
         self._parameter_set = ParameterSet()
         self._representation = representation
 
     @classmethod
-    def from_attributes(
-        cls,
-        name: str,
-        representation: Representation,
-        wires: Wires,
-    ):
+    def _from_attributes(
+        cls, name: str, representation: Representation, wires: Wires
+    ) -> CircuitComponent:
         r"""
         Initializes a circuit component from its attributes (a name, a ``Wires``,
         and a ``Representation``).
 
+        If the Method Resolution Order (MRO) of ``cls`` contains one between ``Ket``, ``DM``,
+        ``Unitary``, and ``Channel``, then the returned component is of that type. Otherwise,
+        it is of type ``CircuitComponent``.
+
+        This function needs to be used with caution, as it does not check that the attributes
+        provided are consistent with the type of the returned component. If used improperly it
+        may initialize, e.g., ``Ket``s with both input and output wires or ``Unitary``s with
+        wires on the bra side.
+
         Args:
             name: The name of this component.
             representation: A representation for this circuit component.
             wires: The wires of this component.
 
         Returns:
-            A circuit component.
+            A circuit component of type ``cls`` with the given attributes.
         """
-        ret = CircuitComponent(name)
-        ret._wires = wires
+        types = {"Ket", "DM", "Unitary", "Channel"}
+        for tp in cls.mro():
+            if tp.__name__ in types:
```

Comment:
should we also check that the wires are compatible with `tp`?

Reply:
ah no, maybe it doesn't matter

Reply:
For how we use it, I think not. It's well documented and hidden by the underscore. But let's have a chat before I merge it